libtrap/src/trap.c

```diff
@@ -1772,48 +1725,32 @@ int trap_ctx_terminate(trap_ctx_t *ctx)
          return trap_errorf(c, TRAP_E_MEMORY, "IFC was not initialized.");
       }
    }
-   return TRAP_E_OK;
+
+   if (c->timeout_thread_initialized == 1) {
+      pthread_cancel(c->timeout_thread);
+   }
+
+   return trap_error(ctx, TRAP_E_OK);
 }
 
 int trap_ctx_recv(trap_ctx_t *ctx, uint32_t ifcidx, const void **data, uint16_t *size)
 {
    int ret_val = 0;
    trap_ctx_priv_t *c = (trap_ctx_priv_t *) ctx;
-   if ((c == NULL) || (c->initialized == 0)) {
+   if (c == NULL || c->initialized == 0) {
       return TRAP_E_NOT_INITIALIZED;
    }
 
-   if (__sync_add_and_fetch(&c->terminated, 0) != 0) {
+   if (c->terminated) {
       return trap_error(c, TRAP_E_TERMINATED);
    }
 
    if (ifcidx >= c->num_ifc_in) {
       return trap_errorf(c, TRAP_E_NOT_SELECTED, "No input ifc to get data from...");
    }
-   if ((c->in_ifc_list[ifcidx].recv != NULL) && (c->in_ifc_list[ifcidx].priv != NULL)) {
-#ifndef DISABLE_BUFFERING
-       /* handle buffering */
-      ret_val = trap_read_from_buffer(c, ifcidx, data, size, c->in_ifc_list[ifcidx].datatimeout);
-      return ret_val;
-#else
-      uint32_t newsize = 0;
-      ret_val = c->in_ifc_list[ifcidx].recv(c->in_ifc_list[ifcidx].priv, c->in_ifc_list[ifcidx].buffer, &newsize, c->in_ifc_list[ifcidx].datatimeout);
-      if (ret_val == TRAP_E_OK) {
-         c->counter_recv_message[ifcidx]++;
-         if (c->in_ifc_list[ifcidx].client_state == FMT_CHANGED) {
-            c->in_ifc_list[ifcidx].client_state = FMT_OK;
-            return TRAP_E_FORMAT_CHANGED;
-         }
-      } else if (ret_val == TRAP_E_FORMAT_MISMATCH) {
-         return ret_val;
-      }
-      (*size) = newsize;
-      (*data) = c->in_ifc_list[ifcidx].buffer;
-      return ret_val;
-#endif
-   } else {
-      return trap_error(c, TRAP_E_NOT_INITIALIZED);
-   }
+
+   ret_val = trap_read_from_buffer(c, ifcidx, data, size, __sync_fetch_and_add(&c->in_ifc_list[ifcidx].datatimeout, 0));
```

Comment:
The problem is not that the value would be old, but that the read is not atomic... meaning that simultaneous read and write of the value might result in situation, where first 4 bytes would contain old data and second 4 bytes might contain new data. Thus resulting into completely different unintentional timeout value.

Reply:
But datatimeout is int32_t, so there are not "first 4 bytes and last 4 bytes". Memory accesses of size native to the architecture should be atomic (if address of the veariable is aligned, which normally is). Unfortunatelly this may not hold for all architectures, so if we want to be 100% protable, we should indeed explicitly assure atomicity.
But isn't there some other macro or something, which ensures atomictiy olny? The `__sync_fetch_and_add` also introduces a memory barrier, which may cause a slowdown, and is not needed in this case.
I don't have time to search for possible solutions, now, I just found there is something in C11 (https://en.cppreference.com/w/c/atomic), but I don't know if we can use this standard (GCC supports stdatomic.h from v4.9, which is from 2014).

Reply:
Don't know why I thought that datatimeout is uint64_t. In that case you are right a the atomic read is most likely overkill.